File: BenchmarkRunner.swift
import Foundation

/// A single scored candidate (for the embedder ranking output).
struct ScoredText: Sendable {
    let text: String
    let score: Float
}

/// Runs head-to-head model benchmarks. Each model is loaded fresh, measured, then freed —
/// independent of the app's live embedding/generation engines, and bounding peak memory.
actor BenchmarkRunner {

    struct EmbedRow: Identifiable, Sendable {
        let id: String          // modelId
        let modelName: String
        let backend: String
        let dim: Int
        let loadMs: Double
        let msPerItem: Double
        let embedsPerSec: Double
        let count: Int
        let top: [ScoredText]
        let failed: Bool
    }

    struct LLMRow: Identifiable, Sendable {
        let id: String
        let modelName: String
        let backend: String
        let loadMs: Double
        let promptTokens: Int
        let genTokens: Int
        let tokensPerSec: Double
        let output: String
        let failed: Bool
    }

    func benchmarkEmbedder(spec: ModelSpec, path: String, backend: LlamaBackend,
                           query: String, candidates: [String], topN: Int = 12) -> EmbedRow {
        let (handle, engaged, loadMs) = load(path: path, embeddings: true, backend: backend, probe: true)
        guard let h = handle else {
            return EmbedRow(id: spec.id, modelName: spec.displayName, backend: "—", dim: 0,
                            loadMs: loadMs, msPerItem: 0, embedsPerSec: 0, count: 0, top: [], failed: true)
        }
        defer { LlmKit.free(h) }

        let pq = prefix(spec.id, isQuery: true)
        let pd = prefix(spec.id, isQuery: false)
        let qv = LlmKit.embed(h, text: pq + query)
        let dim = qv.count

        let t0 = Date()
        var scored: [ScoredText] = []
        for c in candidates where !c.trimmingCharacters(in: .whitespaces).isEmpty {
            let v = LlmKit.embed(h, text: pd + c)
            scored.append(ScoredText(text: c, score: dot(qv, v)))
        }
        let elapsed = Date().timeIntervalSince(t0)
        let count = scored.count
        scored.sort { $0.score > $1.score }

        return EmbedRow(
            id: spec.id, modelName: spec.displayName, backend: engaged.displayName, dim: dim,
            loadMs: loadMs,
            msPerItem: count > 0 ? elapsed * 1000 / Double(count) : 0,
            embedsPerSec: elapsed > 0 ? Double(count) / elapsed : 0,
            count: count,
            top: Array(scored.prefix(topN)),
            failed: false)
    }

    func benchmarkLLM(spec: ModelSpec, path: String, backend: LlamaBackend,
                      system: String, user: String) -> LLMRow {
        let (handle, engaged, loadMs) = load(path: path, embeddings: false, backend: backend, probe: false)
        guard let h = handle else {
            return LLMRow(id: spec.id, modelName: spec.displayName, backend: "—",
                          loadMs: loadMs, promptTokens: 0, genTokens: 0, tokensPerSec: 0, output: "", failed: true)
        }
        defer { LlmKit.free(h) }

        let b = LlmKit.chatBenchmark(h, system: system, user: user, maxTokens: 96, temperature: 0.4, topK: 20)
        return LLMRow(id: spec.id, modelName: spec.displayName, backend: engaged.displayName,
                      loadMs: loadMs, promptTokens: b.promptTokens, genTokens: b.genTokens,
                      tokensPerSec: b.tokensPerSec, output: b.text, failed: b.genTokens == 0)
    }

    // MARK: - Helpers

    private func load(path: String, embeddings: Bool, backend: LlamaBackend, probe: Bool)
        -> (handle: UInt64?, engaged: LlamaBackend, loadMs: Double) {
        let order: [LlamaBackend] = (backend == .cpu) ? [.cpu] : [backend == .auto ? .gpu : backend, .cpu]
        let t0 = Date()
        for b in order {
            guard let h = LlmKit.loadModel(path: path, embeddings: embeddings, backend: b) else { continue }
            if probe {
                let p = LlmKit.embed(h, text: "ok")
                if p.isEmpty || p.contains(where: { $0.isNaN }) { LlmKit.free(h); continue }
            }
            return (h, b, Date().timeIntervalSince(t0) * 1000)
        }
        return (nil, .cpu, Date().timeIntervalSince(t0) * 1000)
    }

    private func prefix(_ id: String, isQuery: Bool) -> String {
        let l = id.lowercased()
        if l.contains("nomic") { return isQuery ? "search_query: " : "search_document: " }
        if l.contains("bge") { return isQuery ? "Represent this sentence for searching relevant passages: " : "" }
        return ""
    }

    private func dot(_ a: [Float], _ b: [Float]) -> Float {
        guard a.count == b.count else { return 0 }
        var s: Float = 0
        for i in 0..<a.count { s += a[i] * b[i] }
        return s
    }
}

File: android/app/src/main/java/com/homeboy/app/ai/BenchmarkRunner.kt
package com.homeboy.app.ai

import com.homeboy.app.ai.ModelRepository.ModelSpec
import com.homeboy.llmkit.LlmKit
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/** A single scored candidate (for the embedder ranking output). */
data class ScoredText(val text: String, val score: Float)

/**
 * Runs head-to-head model benchmarks. Each model is loaded fresh on a SPECIFIC backend
 * (no silent fallback in the request, so NPU vs GPU vs CPU are measured honestly), timed,
 * then freed. Mirrors the iOS `BenchmarkRunner`.
 *
 * All work is native + blocking, so every method hops onto [Dispatchers.Default]. The native
 * layer doesn't expose token counts, so LLM throughput is estimated from output length
 * (`genTokensEstimated = true`) — good enough for a same-device head-to-head comparison.
 */
class BenchmarkRunner(private val nativeLibDir: String) {

    data class EmbedRow(
        val id: String,          // "modelId|backend" — unique per model+backend
        val modelName: String,
        val backend: String,
        val dim: Int,
        val loadMs: Double,
        val msPerItem: Double,
        val embedsPerSec: Double,
        val count: Int,
        val top: List<ScoredText>,
        val failed: Boolean,
        val error: String?
    )

    data class LLMRow(
        val id: String,
        val modelName: String,
        val backend: String,
        val loadMs: Double,
        val genTokens: Int,
        val tokensPerSec: Double,
        /** True when [genTokens]/[tokensPerSec] are estimated from output length, not exact. */
        val genTokensEstimated: Boolean,
        val output: String,
        val failed: Boolean,
        val error: String?
    )

    suspend fun benchmarkEmbedder(
        spec: ModelSpec,
        path: String,
        backend: AiBackend,
        query: String,
        candidates: List<String>,
        topN: Int = 12
    ): EmbedRow = withContext(Dispatchers.Default) {
        val rowId = "${spec.id}|${backend.name}"
        val (handle, loadMs, err) = load(path, embeddings = true, backend = backend, probe = true)
        if (handle == null) {
            return@withContext EmbedRow(
                rowId, spec.displayName, backend.shortLabel, dim = 0, loadMs = loadMs,
                msPerItem = 0.0, embedsPerSec = 0.0, count = 0, top = emptyList(),
                failed = true, error = err
            )
        }
        try {
            val pq = prefix(spec.id, isQuery = true)
            val pd = prefix(spec.id, isQuery = false)
            val qv = LlmKit.embed(handle, pq + query)
            val dim = qv.size

            val t0 = System.nanoTime()
            val scored = ArrayList<ScoredText>(candidates.size)
            for (c in candidates) {
                if (c.isBlank()) continue
                val v = LlmKit.embed(handle, pd + c)
                scored.add(ScoredText(c, dot(qv, v)))
            }
            val elapsedSec = (System.nanoTime() - t0) / 1e9
            val count = scored.size
            scored.sortByDescending { it.score }

            EmbedRow(
                id = rowId, modelName = spec.displayName, backend = backend.shortLabel, dim = dim,
                loadMs = loadMs,
                msPerItem = if (count > 0) elapsedSec * 1000 / count else 0.0,
                embedsPerSec = if (elapsedSec > 0) count / elapsedSec else 0.0,
                count = count,
                top = scored.take(topN),
                failed = false, error = null
            )
        } finally {
            LlmKit.free(handle)
        }
    }

    suspend fun benchmarkLLM(
        spec: ModelSpec,
        path: String,
        backend: AiBackend,
        system: String,
        user: String,
        maxTokens: Int = 96
    ): LLMRow = withContext(Dispatchers.Default) {
        val rowId = "${spec.id}|${backend.name}"
        val (handle, loadMs, err) = load(path, embeddings = false, backend = backend, probe = false)
        if (handle == null) {
            return@withContext LLMRow(
                rowId, spec.displayName, backend.shortLabel, loadMs = loadMs,
                genTokens = 0, tokensPerSec = 0.0, genTokensEstimated = false,
                output = "", failed = true, error = err
            )
        }
        try {
            val t0 = System.nanoTime()
            val raw = runCatching {
                LlmKit.generateChat(handle, system, user, maxTokens = maxTokens, temperature = 0.4f, topK = 20)
            }.getOrNull().orEmpty()
            val genSec = (System.nanoTime() - t0) / 1e9
            val text = stripThink(raw)
            // The native bridge returns only the completion text, not a token count, so estimate
            // it from output length (~4 chars/token for these BPE vocabularies).
            val genTokens = estimateTokens(raw)
            val failed = raw.isBlank()
            LLMRow(
                id = rowId, modelName = spec.displayName, backend = backend.shortLabel,
                loadMs = loadMs,
                genTokens = genTokens,
                tokensPerSec = if (genSec > 0) genTokens / genSec else 0.0,
                genTokensEstimated = true,
                output = text,
                failed = failed,
                error = if (failed) "Loaded but generated no output (decode failed)." else null
            )
        } finally {
            LlmKit.free(handle)
        }
    }

    // MARK: - Helpers

    /**
     * Strict single-backend load with wall-clock timing. Returns a null handle + a diagnostic
     * message on failure. For embedders a [probe] embed guards against a backend that loads but
     * produces NaN/empty vectors (the OpenCL/GPU path does this for BERT-style graphs).
     */
    private fun load(
        path: String,
        embeddings: Boolean,
        backend: AiBackend,
        probe: Boolean
    ): Triple<Long?, Double, String?> {
        val t0 = System.nanoTime()
        val loaded = runCatching {
            LlmKit.setLibraryDir(nativeLibDir)
            LlmKit.loadModel(path, embeddings = embeddings, requested = backend.toKitBackend())
        }.getOrNull()
        val loadMs = (System.nanoTime() - t0) / 1e6
        val handle = loaded?.handle
        if (handle == null || handle == 0L) {
            return Triple(null, loadMs, loadFailureMessage(backend))
        }
        if (probe) {
            val p = runCatching { LlmKit.embed(handle, "ok") }.getOrNull()
            if (p == null || p.isEmpty() || p[0].isNaN()) {
                LlmKit.free(handle)
                return Triple(
                    null, loadMs,
                    "Produced invalid (NaN/empty) embeddings on ${backend.shortLabel} — try CPU or NPU."
                )
            }
        }
        return Triple(handle, loadMs, null)
    }

    private fun loadFailureMessage(backend: AiBackend): String = when (backend) {
        AiBackend.NPU -> "Couldn't load on the NPU — the model may be too large for the Hexagon cDSP, " +
            "or its architecture isn't NPU-supported. Try GPU or CPU."
        AiBackend.GPU -> "Couldn't load on the GPU (Adreno/OpenCL) — try CPU."
        AiBackend.CPU -> "Failed to load — likely an unsupported architecture, the wrong GGUF file " +
            "(e.g. an mmproj projector or one shard of a split model), or out of memory."
    }

    /** Rough token count from output length; matches the ~4 chars/token BPE ratio. */
    private fun estimateTokens(text: String): Int {
        if (text.isBlank()) return 0
        return kotlin.math.ceil(text.length / 4.0).toInt().coerceAtLeast(1)
    }

    private fun stripThink(s: String): String {
        var out = s
        while (true) {
            val start = out.indexOf("<think>")
            if (start < 0) break
            val end = out.indexOf("</think>", start)
            out = if (end < 0) out.substring(0, start)
            else out.substring(0, start) + out.substring(end + "</think>".length)
        }
        return out.trim()
    }

    /** Instruction prefix each embedding family needs, on the query vs document side. */
    private fun prefix(id: String, isQuery: Boolean): String {
        val l = id.lowercase()
        return when {
            l.contains("nomic") -> if (isQuery) "search_query: " else "search_document: "
            l.contains("bge") -> if (isQuery) "Represent this sentence for searching relevant passages: " else ""
            else -> ""
        }
    }

    /** Dot product of two L2-normalized vectors (= cosine similarity). */
    private fun dot(a: FloatArray, b: FloatArray): Float {
        if (a.size != b.size) return 0f
        var s = 0f
        for (i in a.indices) s += a[i] * b[i]
        return s
    }
}
